### ISSUE TITLE
3.1.3: FormType: the translator doesn't work in other methods than buildForm()

### ISSUE THREAD

**Issue by enrico69** [symfony/symfony#19789]
Symfony version 3.1.3
PHP 5.5.9
Ubuntu

I have an issue to translate "on the fly" some elements in a form class. This is very strange because the translator works perfectly for the labels of each form (in the _buildForm()_ method) but not in the privates ones.

According to the profiler, the translator is not even called in these methods.

Here the translator is called and works:

```php
  public function buildForm(FormBuilderInterface $builder, array $options) {
    $this->options = $options;
    $builder
        ->add('title', TextType::class, array(
                'label' => $this->getTranslator()->trans('Title')
            )
        )
```

But here, in the same class, it doesn't:

```php
  private function getMaterialList() {
    try {
        $arrResult = $this->em->getRepository('AppBundle:Material')->findAllOrderedByNameWp();
        foreach($arrResult as $material) {
            $material->setName($this->getTranslator()->trans($material->getName()));
        }
    } catch (NoResultException $e) {
        unset($e);
    } catch (Exception $ex) {
        $this->logger->info('Unable to list materials from booktype class ' . $ex->getMessage());
        throw new \Exception(
            $this->getTranslator()->trans('errorListingMaterials')
        );
    }
```

**Comment by HeahDude** [comment]
Very strange. Are you sure `getMaterialList` is called?

**Comment by enrico69** [comment]
Yes it is called because I see the results in the choice list. I can't figure out where this problem comes from.

Below is the complete class:

```
<?php
namespace AppBundle\Form;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\Form\Extension\Core\Type\TextareaType;
use Symfony\Bridge\Doctrine\Form\Type\EntityType;
use Symfony\Component\Form\Extension\Core\Type\SubmitType;
use Symfony\Component\Translation\Translator;
use Doctrine\Bundle\DoctrineBundle\Registry as Doctrine;
use Symfony\Component\HttpFoundation\RequestStack;
use Doctrine\ORM\NoResultException;

/**
 * This class represents the form for the book entity
 */
class BookType extends AbstractType {
    
    private $translator;
    private $em;
    private $request;
    private $logger;
    
    private function getTranslator() {
        
        if(!isset($this->translator)) {
            $this->translator = new Translator($this->options['locale']);
        }
        
        return $this->translator;
    }
    
    /**
     * List the Editors
     * 
     * @return an array of Editors
     * @throws \Exception
     */
    private function getEditorList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Editor')->findAllOrderedByNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list editors from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingEditors')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Materials
     * 
     * @return an array of materials
     * @throws \Exception
     */
    private function getMaterialList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Material')->findAllOrderedByNameWp();
            foreach($arrResult as $material) {
                $material->setName($this->getTranslator()->trans($material->getName()));
            }
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        } catch (Exception $ex) {
            $this->logger->info('Unable to list materials from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingMaterials')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Formats
     * 
     * @return an array of Formats
     * @throws \Exception
     */
    private function getFormatList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Format')->findAllOrderedByNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list formats from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingFormats')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Languages
     * 
     * @return an array of Languages
     * @throws \Exception
     */
    private function getLanguageList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Language')->findAllOrderedByNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list languages from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingLanguages')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Locations
     * 
     * @return an array of Locations
     * @throws \Exception
     */
    private function getLocationList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Location')->findAllOrderedByNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list locations from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingLocations')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Authors
     * 
     * @return an array of Authors
     * @throws \Exception
     */
    private function getAuthorList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Author')->findAllOrderedBySurNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list authors from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingAuthors')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * List the Locations
     * 
     * @return an array of Kinds
     * @throws \Exception
     */
    private function getKindList() {
        try {
            $arrResult = $this->em->getRepository('AppBundle:Kind')->findAllOrderedByNameWp();
        } catch (NoResultException $e) {
            $arrResult = array();
            unset($e);
        }catch (Exception $ex) {
            $this->logger->info('Unable to list kinds from booktype class ' . $ex->getMessage());
            throw new \Exception(
                $this->getTranslator()->trans('errorListingKinds')
            );
        }
        
        return $arrResult;
    }
    
    /**
     * Constructor
     * 
     * @param Doctrine $doctrine
     * @param RequestStack $requestStack is the request stacl
     */
    public function __construct(Doctrine $doctrine, RequestStack $requestStack, $logger) {
        $this->em = $doctrine->getManager();
        $this->request = $requestStack;
        $this->logger = $logger;        
    }
    
    public function buildForm(FormBuilderInterface $builder, array $options) {

        $this->options = $options;
        $builder
            ->add('title', TextType::class, array(
                    'label' => $this->getTranslator()->trans('Title')
                )
            )
            ->add('description', TextareaType::class, array(
                    'label' => $this->getTranslator()->trans('Description')
                )
            )
            ->add('year', TextType::class, array(
                    'label' => $this->getTranslator()->trans('Year')
                )
            )
            ->add('editor', EntityType::class, array(
                    'class' => 'AppBundle:Editor',
                    'label' => $this->getTranslator()->trans('Editor.single'),
                    'choices' => $this->getEditorList()
                )
            )
            ->add('format', EntityType::class, array(
                    'class' => 'AppBundle:Format',
                    'label' => $this->getTranslator()->trans('Format.single'),
                    'choices' => $this->getFormatList()
                )
            )
            ->add('material', EntityType::class, array(
                    'class' => 'AppBundle:Material',
                    'label' => $this->getTranslator()->trans('Material'),
                    'choices' => $this->getMaterialList()
                )
            )
            ->add('language', EntityType::class, array(
                    'class' => 'AppBundle:Language',
                    'label' => $this->getTranslator()->trans('Language.single'),
                    'choices' => $this->getLanguageList()
                )
            )
            ->add('location', EntityType::class, array(
                    'class' => 'AppBundle:Location',
                    'label' => $this->getTranslator()->trans('Location.plural'),
                    'choices' => $this->getLocationList(),
                    'multiple' => true
                )
            )
            ->add('author', EntityType::class, array(
                    'class' => 'AppBundle:Author',
                    'label' => $this->getTranslator()->trans('Author.plural'),
                    'choices' => $this->getAuthorList(),
                    'multiple' => true
                )
            )
            ->add('kind', EntityType::class, array(
                    'class' => 'AppBundle:Kind',
                    'label' => $this->getTranslator()->trans('Kind.plural'),
                    'choices' => $this->getKindList(),
                    'multiple' => true
                )
            )
            ->add('isbn', TextType::class, array(
                    'label' => $this->getTranslator()->trans('ISBN'),
                    'required' => false
                )
            )
            ->add('keywords', TextType::class, array(
                    'label' => $this->getTranslator()->trans('Keywords')
                )
            )
            ->add('slug', TextType::class, array(
                    'label' => $this->getTranslator()->trans('Slug')
                )
            )  
            ->add('submit', SubmitType::class, array(
                'label' => $this->getTranslator()->trans('Save')
            ));
    }
    
    public function configureOptions(OptionsResolver $resolver) {
        $resolver->setDefaults(array(
                'data_class' => 'AppBundle\Entity\Book',
                'locale' => 'en', // the default locale
            )
        );
    }
    
}
```

**Comment by xabbuh** [comment]
Calls to the translator you created inside your form type will never appear in the profiler. If you want to see them in the profiler, you should rather inject the core translator into your form type (you therefore need to register it as a service). However, I actually do not understand why you are dealing with translating the labels manually at all. Why don't you let the Form component deal with that task?

**Comment by enrico69** [comment]
> Why don't you let the Form component deal with that task?

As I am new with internationalization with Symfony, I was not aware that the component form was doing it automatically. Thanks for that. I corrected this, lightweighting the code.

Regarding my problem, despite walking through the code of the translator, I didn't manage to understand why the translation process was not working properly. So I came back to your solution: injecting the service in the form.
